src/OpenIddict.Core/Infrastructure/OpenIddictProvider.Serialization.cs

```diff
@@ -5,20 +5,55 @@
  */
 
 using System;
+using System.Diagnostics;
+using System.Security.Claims;
 using System.Threading.Tasks;
 using AspNet.Security.OpenIdConnect.Extensions;
 using AspNet.Security.OpenIdConnect.Server;
 using JetBrains.Annotations;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
 
 namespace OpenIddict.Infrastructure {
     public partial class OpenIddictProvider<TUser, TApplication, TAuthorization, TScope, TToken> : OpenIdConnectServerProvider
         where TUser : class where TApplication : class where TAuthorization : class where TScope : class where TToken : class {
         public override async Task SerializeRefreshToken([NotNull] SerializeRefreshTokenContext context) {
             var services = context.HttpContext.RequestServices.GetRequiredService<OpenIddictServices<TUser, TApplication, TAuthorization, TScope, TToken>>();
 
-            // Persist a new token entry in the database.
-            var identifier = await services.Tokens.CreateAsync(OpenIdConnectConstants.TokenTypeHints.RefreshToken);
+            Debug.Assert(context.Request.RequestType == OpenIdConnectRequestType.TokenRequest,
+                "The request should be a token request.");
+
+            Debug.Assert(!context.Request.IsClientCredentialsGrantType(),
+                "A refresh token should not be issued when using grant_type=client_credentials.");
+
+            string identifier;
+
+            var user = await services.Users.FindByIdAsync(context.Ticket.Principal.GetClaim(ClaimTypes.NameIdentifier));
+            if (user == null) {
+                throw new InvalidOperationException("The user cannot be retrieved from the database.");
+            }
+
+            // If the client application sending the token request is known,
+            // ensure the token is attached to the corresponding client entity.
```

Comment:
My bad, I miss that, we associate the same token with application and user as well, so forget my prev comment, EF handles it just well